tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lower_alloc_dealloc.cc

```diff
@@ -0,0 +1,203 @@
+                                          Value allocated_byte_ptr,
+                                          ArrayRef<Value> sizes) const {
+    auto memref_desc = MemRefDescriptor::undef(
+        rewriter, loc, typeConverter->convertType(memref_type));
+
+    // TF AllocateRaw returns aligned pointer => AllocatedPtr == AlignedPtr.
+    Value allocated_type_ptr = rewriter.create<LLVM::BitcastOp>(
+        loc, getElementPtrType(memref_type), allocated_byte_ptr);
+    memref_desc.setAllocatedPtr(rewriter, loc, allocated_type_ptr);
+    memref_desc.setAlignedPtr(rewriter, loc, allocated_type_ptr);
+    memref_desc.setConstantOffset(rewriter, loc, 0);
+
+    if (memref_type.getRank() == 0) {
+      return memref_desc;
+    }
+
+    // Compute strides and populate descriptor `size` and `stride` fields.
+    Value stride_carried = createIndexConstant(rewriter, loc, 1);
+    for (int pos = sizes.size() - 1; pos >= 0; --pos) {
+      Value size = sizes[pos];
+      memref_desc.setSize(rewriter, loc, pos, size);
+      memref_desc.setStride(rewriter, loc, pos, stride_carried);
+      // Update stride
+      if (pos > 0) {
+        stride_carried =
+            rewriter.create<LLVM::MulOp>(loc, stride_carried, size);
+      }
+    }
+    return memref_desc;
+  }
+  LogicalResult matchAndRewrite(
+      memref::AllocOp alloc_op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const override;
+};
+
+// Emits LLVM IR to malloc a device memory.
+LogicalResult ConvertMemRefAllocOpToDispatchOpPattern::matchAndRewrite(
+    memref::AllocOp alloc_op, ArrayRef<Value> operands,
+    ConversionPatternRewriter& rewriter) const {
+  mlir::Operation* op = alloc_op.getOperation();
+  ModuleOp module = op->getParentOfType<ModuleOp>();
+
+  Location loc = op->getLoc();
+
+  // check address space
+  auto memref = alloc_op.getResult();
+  MemRefType memref_type = memref.getType().cast<MemRefType>();
+  Attribute memorySpace = memref_type.getMemorySpace();
+  if (!memorySpace) {
+    return failure();
+  }
+  if (!memorySpace.isa<StringAttr>()) {
+    return failure();
+  }
+
+  // get ral context
+  FuncOp std_func_op = alloc_op->getParentOfType<FuncOp>();
+  LLVM::LLVMFuncOp llvm_func_op = alloc_op->getParentOfType<LLVM::LLVMFuncOp>();
+  if (!std_func_op && !llvm_func_op) {
+    return failure();
+  }
+  Value context_arg;
+  if (std_func_op) {
+    context_arg = std_func_op.getArgument(0);
+  } else {
+    context_arg = llvm_func_op.getArgument(0);
+  }
+
+  // Set all dynamic sizes to 1 and compute fake strides.
+  SmallVector<Value, 4> dyn_sizes(memref_type.getNumDynamicDims(),
+                                  createIndexConstant(rewriter, loc, 1));
+  // Get memref descriptor sizes.
+  SmallVector<Value, 4> sizes;
+  SmallVector<Value, 4> strides;
+  Value sizeBytes;
+  getMemRefDescriptorSizes(loc, memref_type, dyn_sizes, rewriter, sizes,
+                           strides, sizeBytes);
```

Comment:
Can you improve the documentation here? Maybe with a sample of IR (or pseudo code) about what we intend to generate?

Maybe if you document the pattern class itself with such IR samples it would be easier to follow here.

Reply:
Add doc for ConvertMemRefAllocOpToDispatchOpPattern and ConvertMemRefDeallocOpToDispatchOpPattern